In [78]:
from surprise import Dataset
from surprise.model_selection import cross_validate, GridSearchCV
from surprise import SVD, SVDpp, NMF, KNNBasic
from surprise.accuracy import rmse, mae

In [79]:
data = Dataset.load_builtin('ml-100k')

In [80]:
models = [
    {'algo': SVD, 'param_grid': {'n_epochs': [5, 10], 'lr_all': [0.002, 0.005], 'reg_all': [0.4, 0.6]}},
    {'algo': SVDpp, 'param_grid': {'n_epochs': [5, 10], 'lr_all': [0.002, 0.005], 'reg_all': [0.4, 0.6]}},
    {'algo': NMF, 'param_grid': {'n_factors': [10, 20], 'n_epochs': [10, 20]}}
]

In [82]:
for model in models:
    algo = model['algo']
    param_grid = model['param_grid']
    result = GridSearchCV(algo, param_grid, measures=['rmse', 'mae'], cv=3)
    result.fit(data)
    print(f'Algorithm: {algo.__name__}')
    print(f'Best parameters for RMSE: {result.best_params['rmse']}')
    print(f'RMSE: {result.best_score['rmse']}')
    print(f'Best parameters for MAE: {result.best_params['mae']}')
    print(f'MAE: {result.best_score['mae']}')
    print('-------------------------------------------------------')

Algorithm: SVD
Best parameters for RMSE: {'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.4}
RMSE: 0.9643974185029288
Best parameters for MAE: {'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.4}
MAE: 0.7731227818402918
-------------------------------------------------------
Algorithm: SVDpp
Best parameters for RMSE: {'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.4}
RMSE: 0.9643362720714027
Best parameters for MAE: {'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.4}
MAE: 0.772847758813581
-------------------------------------------------------
Algorithm: NMF
Best parameters for RMSE: {'n_factors': 10, 'n_epochs': 20}
RMSE: 1.0646028456001226
Best parameters for MAE: {'n_factors': 20, 'n_epochs': 20}
MAE: 0.8431135055575524
-------------------------------------------------------


**SVD and SVDpp:** <br>
The best parameters for both algorithms are the same: n_epochs: 10, lr_all: 0.005, reg_all: 0.4.<br>
The RMSE for both algorithms is about 0.964, and the MAE is about 0.773.<br>
SVDpp takes longer to execute than SVD. This is due to the greater complexity of the algorithm itself. The choice between algorithms depends on the specific needs of the project: if speed is critical, SVD may be the optimal choice, otherwise, if maximum accuracy is the main goal, SVDpp can be considered.<br>

**NMF:**<br>
Best settings: n_factors: 10, n_epochs: 20.<br>
The RMSE is about 1.065 and the MAE is about 0.843, an experiment with other settings can get better results.<br>
The optimal n_factors parameter was found to be different for different metrics (RMSE and MAE), showing that the metrics take into account different aspects of recommendation quality. RMSE focuses more on the accuracy of forecasts, while MAE estimates the mean absolute error. It is important to understand which metric is more important to use and take this into account when choosing parameters.<br>